components/api/src/main/java/com/hotels/styx/api/messages/StreamingHttpRequest.java

```diff
@@ -0,0 +1,693 @@
+    public static Builder post(String uri, Observable<ByteBuf> body) {
+        return new Builder(POST, uri).body(body);
+    }
+
+    /**
+     * Creates a request with the PUT method.
+     *
+     * @param uri URI
+     * @param body body
+     * @param  body type
+     * @return {@code this}
+     */
+    public static Builder put(String uri, Observable<ByteBuf> body) {
+        return new Builder(PUT, uri).body(body);
+    }
+
+    /**
+     * Creates a request with the PATCH method.
+     *
+     * @param uri URI
+     * @param body body
+     * @param  body type
+     * @return {@code this}
+     */
+    public static Builder patch(String uri, Observable<ByteBuf> body) {
+        return new Builder(PATCH, uri).body(body);
+    }
+
+    @Override
+    public HttpVersion version() {
+        return this.version;
+    }
+
+    @Override
+    public HttpHeaders headers() {
+        return headers;
+    }
+
+    @Override
+    public List<HttpCookie> cookies() {
+        return cookies;
+    }
+
+    @Override
+    public List<String> headers(CharSequence name) {
+        return headers.getAll(name);
+    }
+
+    @Override
+    public Observable<ByteBuf> body() {
+        return body;
+    }
+
+    /**
+     * Gets the unique ID for this request.
+     *
+     * @return request ID
+     */
+    public Object id() {
+        return id;
+    }
+
+    /**
+     * Returns the {@link HttpMethod} of this.
+     *
+     * @return the HTTP method
+     */
+    public HttpMethod method() {
+        return method;
+    }
+
+    /**
+     * Returns the requested URI (or alternatively, path).
+     *
+     * @return The URI being requested
+     */
+    public Url url() {
+        return url;
+    }
+
+    /**
+     * Returns the requested path.
+     *
+     * @return the path being requested
+     */
+    public String path() {
+        return url.path();
+    }
+
+    /**
+     * Returns {@code true} if and only if the connection can remain open and thus 'kept alive'.
+     * This methods respects the value of the {@code "Connection"} header first and if this has no such header
+     * then the keep-alive status is determined by the HTTP version, that is, HTTP/1.1 is keep-alive by default,
+     * HTTP/1.0 is not keep-alive by default.
+     *
+     * @return true if the connection is keep-alive
+     */
+    public boolean keepAlive() {
+        return HttpMessageSupport.keepAlive(headers, version);
+    }
+
+    /**
+     * Checks if the request has been transferred over a secure connection. If the protocol is HTTPS and the
+     * content is delivered over SSL then the request is considered to be secure.
+     *
+     * @return true if the request is transferred securely
+     */
+    public boolean isSecure() {
+        return secure;
+    }
+
+    /**
+     * Returns the remote client address that initiated the current request.
+     *
+     * @return the client address for this request
+     */
+    public InetSocketAddress clientAddress() {
```

Comment:
Client address is conceptually not part of HTTP request/response objects. Therefore we should not have it our HTTP request abstraction. Please remove this. 

I recognise it appears in the existing HttpRequest/Response objects. So for the sake of backwards compatibility we can keep it as a private filed so the requests/responses can be converted between the formats without losing information. But nevertheless we should not have this as a part of public API.

The clientAddress should be exposed to the interceptors as a part of interceptor context instead.